Fix: issue 629

aaryankotharii opened the pull request:
##### This PR migrates regions storage from UserDefaults to FileManager. Same file structure followed as mentioned in issue description. 

<hr>
fixes issue: #629 
<hr>

### Code Changes

- Created `FileManagerProtocol` to abstract FileManager operations for testability.
- Updated RegionService to store and fetch defaultRegion and custom Regions in File system. 
- Created `FileManagerMock` and updated `RegionsServiceTests`.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/OneBusAway/onebusaway-ios?pullRequest=714) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/OneBusAway/onebusaway-ios?pullRequest=714) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OneBusAway/onebusaway-ios?pullRequest=714) it.</sub>

aaronbrethorst commented:
Thank you @aaryankotharii! I'll review and get back to you soon!

ualch9 reviewed (changes requested):
Thanks for this PR, @aaryankotharii, we really appreciate it! Great job with updating the tests to ensure quality! I have a few concerns with the layering and the addition of more state.

Also, please test that users who update are able to continue using the app with no interruptions. For example, regions are currently stored in UserDefaults and this PR changes it to storing on disk. Please run the app in a current App Store version, then run a version of the app with this PR and validate that the user can continuing the app without it being confused on the currently selected region. 

Please ask questions if you have any, or are unclear about my comments!

aaryankotharii commented:
Hello @ualch9, I have made the following changes as requested above.
1. replaced `PropertyListEncoder / Decoder` with `JSONEncoder / Decoder`.
2. replaced enum `FileDestination` logic with simple static constants in `RegionsService`.
3. renamed `FileManagerProtocol` to `RegionsServiceFileManagerProtocol`.

ualch9 commented:
Thank you @aaryankotharii! Apologies for the delayed turnaround in reviewing, I'll get to this ASAP.

aaryankotharii commented:
no worries @ualch9, but there are a few migration issues I'm not sure how to address. for example, how to approach fetching custom regions, of existing users, which are stored in userdefaults ?

ualch9 commented:
> no worries @ualch9, but there are a few migration issues I'm not sure how to address. for example, how to approach fetching custom regions, of existing users, which are stored in userdefaults ?

Excellent question... @aaronbrethorst, do you have any insight on if custom regions are used enough for us to support a migration path?

aaronbrethorst commented:
I would not worry about custom regions at this time. Thanks for checking!

ualch9 commented on OBAKitCore/Location/Regions/RegionsService.swift:
```diff
@@ -204,17 +207,10 @@ public class RegionsService: NSObject, LocationServiceDelegate {
 
     // MARK: - Custom Regions
     /// Adds the provided custom region to the RegionsService.
-    /// If an existing custom region with the same `regionIdentifier` exists, the new region replaces the existing region.
```
With this new file manager service, what happens if an existing custom region with the same `regionIdentifier` exists?

ualch9 commented on OBAKitCore/Location/Regions/RegionsService.swift:
```diff
@@ -262,16 +257,20 @@ public class RegionsService: NSObject, LocationServiceDelegate {
     static let currentRegionUserDefaultsKey = "OBACurrentRegionUserDefaultsKey"
     static let regionsUpdatedAtUserDefaultsKey = "OBARegionsUpdatedAtUserDefaultsKey"
+    static let defaultRegionsPath = URL.applicationSupportDirectory.appendingPathComponent("default-regions.json")
+    static let customRegionsPath = URL.documentsDirectory.appendingPathComponent("Regions/custom-regions")
```
👍

ualch9 commented on OBAKitTests/Location/RegionsServiceTests.swift:
```diff
@@ -68,10 +70,9 @@ class RegionsServiceTests: OBATestCase {
         stubRegions(dataLoader: dataLoader)
```
Please update the function signature comment to `saved to disk` instead of `user defaults`.

ualch9 commented on OBAKitTests/Location/RegionsServiceTests.swift:
```diff
@@ -83,26 +84,25 @@ class RegionsServiceTests: OBATestCase {
         stubRegions(dataLoader: dataLoader)
```
Please update the function signature comment to `saved to disk` instead of `user defaults`.

ualch9 reviewed (commented):
Just a documentation nit and a question for you to answer. Thanks!

aaronbrethorst commented:
@aaryankotharii - need anything from me in order to get this wrapped up?
